ReadMe for out of process samples (#470)
Azure/azure-functions-sql-extension: docs/BindingsGuide_DotnetOutOfProc.md

> # SQL Extension .NET Worker

Reviewer: ```suggestion
# # Azure SQL bindings for Azure Functions - .NET (Isolated Process)
```

> Welcome to the Sql Extension .NET Worker Repository. The .NET Worker provides .NET 6 support for SQL Bindings in Azure Functions, introducing an **Isolated Model**, running as an out-of-process language worker that is separate from the Azure Functions runtime. This allows you to have full control over your application's dependencies as well as other new features like a middleware pipeline. A .NET Isolated function app works differently than a .NET Core 3.1 function app. For .NET Isolated, you build an executable that imports the .NET Isolated language worker as a NuGet package. Your app includes a [`Program.cs`](https://github.com/Azure/azure-functions-sql-extension/blob/main/samples/samples-outofproc/Program.cs) that starts the worker.

Reviewer: You can move this blurb back into a README in the out of proc directory. This guide should be specifically for creating functions with SQL bindings in the different languages - where as this stuff is more about the actual worker itself. 

(same goes for the other content in here that isn't relevant to creating a function)

## Binding Model

.NET Isolated introduces a new binding model, slightly different from the binding model exposed in .NET Core 3 Azure Functions. More information can be [found here](https://github.com/Azure/azure-functions-dotnet-worker/wiki/.NET-Worker-bindings). Please review our samples for usage information.

## Samples

You can find samples on how to use Sql Extension .NET Worker under `samples\samples-outofproc` ([link](https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-outofproc)).

## Create and run .NET Isolated functions

**Note: Visual Studio and Visual Studio Code support is on the way. In the meantime, please use `azure-functions-core-tools` or the sample projects as a starting point.**  

### Install .NET 6.0

Download .NET 6.0 [from here](https://dotnet.microsoft.com/download/dotnet/6.0)

### Install the Azure Functions Core Tools

To download Core Tools, please check out our docs at [Azure Functions Core Tools](https://github.com/Azure/azure-functions-core-tools)

### Create an Isolated Function App

Run the below in the Terminal/CLI or with PowerShell.

    ```
    mkdir MyApp
    cd MyApp

>     func init --worker-runtime dotnet (Isolated Process)

Reviewer: ```suggestion
    func init --worker-runtime dotnetIsolated
```

    ```

   or in an empty directory, run `func init` and select `dotnet (Isolated Process)`

> ### Add a function: 

Reviewer: Remove this - adding functions isn't special to out of proc so we can just link to the normal .NET guide at the end of this

### Configure Function App

1. Get your SQL connection string
    
    Local SQL Server - Use this connection string, replacing the placeholder values for the database and password.  
      
    `Server=localhost;Initial Catalog={db_name};Persist Security Info=False;User ID=sa;Password={your_password};` Azure SQL Server - Browse to the SQL Database resource in the [Azure portal](https://ms.portal.azure.com/)  
    \- In the left blade click on the **Connection Strings** tab  
    \- Copy the **SQL Authentication** connection string  
      
    (_Note: when pasting in the connection string, you will need to replace part of the connection string where it says '{your\_password}' with your Azure SQL Server password_)
2. Open the generated `local.settings.json` file and in the `Values` section verify you have the below. If not, add the below and replace `{connection_string}` with the your connection string from the previous step:
    
    ```
    "AzureWebJobsStorage": "UseDevelopmentStorage=true",
    "FUNCTIONS_WORKER_RUNTIME": "dotnet-isolated",
    "SqlConnectionString": "{connection_string}"
    
    ```
    

> 3. Verify your `host.json` looks like the below:

Reviewer: Do we need this? There isn't anything special in here that we need to modify so I'd say just get rid of this and let the default generated be used

>  # Azure SQL bindings for Azure Functions - .NET (Isolated Process)
>  ## Binding Model

Reviewer: ```suggestion
 # Azure SQL bindings for Azure Functions - .NET (Isolated Process)
 
 ## Binding Model
```

- **ConnectionStringSetting**: Specifies the name of the app setting that contains the SQL connection string used to connect to a database. The connection string must follow the format specified [here](https://docs.microsoft.com/dotnet/api/microsoft.data.sqlclient.sqlconnection.connectionstring?view=sqlclient-dotnet-core-2.0).

The following are valid binding types for the result of the query/stored procedure execution:

- **IEnumerable&lt;T&gt;**: Each element is a row of the result represented by `T`, where `T` is a user-defined POCO, or Plain Old C# Object. `T` should follow the structure of a row in the queried table. See the [Query String](#query-string) section for an example of what `T` should look like.
- **IAsyncEnumerable&lt;T&gt;**: Each element is again a row of the result represented by `T`, but the rows are retrieved "lazily". A row of the result is only retrieved when `MoveNextAsync` is called on the enumerator. This is useful in the case that the query can return a very large amount of rows.

> - **String**: A JSON string representation of the rows of the result (an example is provided [here](https://github.com/Azure/azure-functions-sql-extension/blob/main/samples/samples-csharp/InputBindingSamples/GetProductsString.cs)).

Reviewer: ```suggestion
- **String**: A JSON string representation of the rows of the result (an example is provided [here](https://github.com/Azure/azure-functions-sql-extension/blob/main/samples/samples-outofproc/InputBindingSamples/GetProductsString.cs)).
```

    public string Name { get; set; }

    public int Cost { get; set; }

}
```

#### Empty Parameter Value

In this case, the parameter value of the `@Name` parameter is an empty string.

```csharp
  [Function("GetProductsNameEmpty")]
  public static IEnumerable<Product> Run(
      [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "getproducts-nameempty/{cost}")]
      HttpRequest req,
      [SqlInput("select * from Products where Cost = @Cost and Name = @Name",
          CommandType = System.Data.CommandType.Text,
          Parameters = "@Cost={cost},@Name=",
          ConnectionStringSetting = "SqlConnectionString")]
      IEnumerable<Product> products)
  {
      return products;
  }
  ```

#### Null Parameter Value

If the `{name}` specified in the `getproducts-namenull/{name}` URL is "null", the query returns all rows for which the Name column is `NULL`. Otherwise, it returns all rows for which the value of the Name column matches the string passed in `{name}`

```csharp
  [Function("GetProductsNameNull")]
  public static IEnumerable<Product> Run(
      [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "getproducts-namenull/{name}")]
      HttpRequest req,

>       [Sql("if @Name is null select * from Products where Name is null else select * from Products where @Name = name",

Reviewer: ```suggestion
      [SqlInput("if @Name is null select * from Products where Name is null else select * from Products where @Name = name",
```

The following are valid binding types for the rows to be upserted into the table:

Each element is a row represented by `T`, where `T` is a user-defined POCO, or Plain Old C# Object. `T` should follow the structure of a row in the queried table. See the [Query String](#query-string) for an example of what `T` should look like.
- **T**: Used when just one row is to be upserted into the table.
- **T[]**: Each element is again a row of the result represented by `T`. This output binding type requires manual instantiation of the array in the function.
**Note**: As stated in the functions [documentation](https://learn.microsoft.com/azure/azure-functions/dotnet-isolated-process-guide#output-bindings)
- Because .NET isolated projects run in a separate worker process, bindings can't take advantage of rich binding classes, such as ICollector<T>, IAsyncCollector<T>, and CloudBlockBlob.

> The repo contains examples of each of these binding types [here](https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-csharp/OutputBindingSamples). A few examples are also included [below](#samples-for-output-bindings).

Reviewer: ```suggestion
The repo contains examples of each of these binding types [here](https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-outofproc/OutputBindingSamples). A few examples are also included [below](#samples-for-output-bindings).
```

- Press 'F1' and search for 'Azure Functions: Create Function'
- Choose HttpTrigger ->  (Provide a function name) -> Company.namespace is fine -> anonymous
- In the file that opens, replace the `public static async Task<IActionResult> Run` block with the below code

    ```csharp
    [Function("AddEmployees")]
    [SqlOutput("dbo.Employees", ConnectionStringSetting = "SqlConnectionString")]
    public static Employee[] Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "addemployees-array")]
        HttpRequestData req)
    {
        Employee[] output = new Employee[]
            {
                new Employee
                {
                    EmployeeId = 1,
                    FirstName = "Hello",
                    LastName = "World",
                    Company = "Microsoft",
                    Team = "Functions"
                },
                new Employee
                {
                    EmployeeId = 2,
                    FirstName = "Hi",
                    LastName = "SQLupdate",
                    Company = "Microsoft",
                    Team = "Functions"
                },
            };

        return output;
    }
    ```

>     *In the above, "dbo.Employees" is the name of the table our output binding is upserting into. The line below is similar to the input binding and specifies where our SqlConnectionString is. For more information on this, see the [SqlAttribute for Output Bindings](#sqloutputattribute-for-output-bindings) section*

Reviewer: ```suggestion
    *In the above, "dbo.Employees" is the name of the table our output binding is upserting into. The line below is similar to the input binding and specifies where our SqlConnectionString is. For more information on this, see the [SqlOutputAttribute for Output Bindings](#sqloutputattribute-for-output-bindings) section*
```

#### Columns with Default Values
In the case where one of the primary key columns has a default value, there are also two options based on how the function defines the output object:
1. If the column with a default value is not included in the output object, then a straight insert is always performed with the other values. See [AddProductWithDefaultPK](../samples/samples-outofproc/OutputBindingSamples/AddProductWithDefaultPK.cs) for an example.
2. If the column with a default value is included then a merge is performed similar to what happens when no default column is present. If there is a nullable column with a default value, then the provided column value in the output object will be upserted even if it is null.

## Key differences with .NET (Isolated Process)

> Please refer to the functions documentation [here](https://learn.microsoft.com/azure/azure-functions/dotnet-isolated-process-guide#output-bindings)

Reviewer: ```suggestion
Please refer to the functions documentation [here](https://learn.microsoft.com/azure/azure-functions/dotnet-isolated-in-process-differences)
```
